bodyswap.lua

```diff
@@ -0,0 +1,255 @@
+--[====[
+
+bodyswap
+========
+
+Shifts player control over to another unit (including wild animals) in adventure mode.
+Usage:
+
+:bodyswap:
+    Swaps into the unit selected in the UI (for example, when viewing the unit's status screen or description).
+
+:bodyswap -unit [unitId]:
+    Swaps into the unit corresponding to the specified id.
+
+]====]
+
+local utils = require 'utils'
+validArgs = validArgs or utils.invert({
+'unit',
+'help'
+})
+local args = utils.processArgs({...}, validArgs)
+
+local usage = [====[
+
+bodyswap
+========
+This script allows the player to gain control over a new unit in adventurer mode
+whilst simultaneously loosing control over their current character.
+
+To specify the target unit, simply select it in the user interface,
+such as by opening the unit's status screen or viewing its description
+and enter "bodyswap" in the DFHack console.
+
+Alternatively, the target unit can be specified by its unit id as shown below.
+
+Arguments::
+
+    -unit id
+        replace "id" with the unit id of your target
+        example: 
+            bodyswap -unit 42
+
+]====]
+
+if args.help then
+ print(usage)
+ return
+end
+
+if not dfhack.world.isAdventureMode() then
+  qerror("This script can only be used in adventure mode!")
+end
+
+function setNewAdvNemFlags(nem)
+  nem.flags.ACTIVE_ADVENTURER = true
+  nem.flags.RETIRED_ADVENTURER = false
+  nem.flags.ADVENTURER = true
+end
+function setOldAdvNemFlags(nem)
+  nem.flags.ACTIVE_ADVENTURER = false
+  nem.flags.RETIRED_ADVENTURER = true
+  nem.unit.idle_area.x = nem.unit.pos.x
+  nem.unit.idle_area.y = nem.unit.pos.y
+  nem.unit.idle_area.z = nem.unit.pos.z
+end
+
+function clearNemesisFromSite(nem)
+-- this is a workaround for a bug which tends to cause duplication of the unit entry in df.global.world.units.active when the site to which a historical figure is linked is reloaded with the unit present
+-- appears to fix the problem without causing any noticeable issues
+  if not nem.figure then
+    return
+  end
+  for _,link in ipairs(nem.figure.site_links) do
+    local site = df.world_site.find(link.site)
+    for i = #site.unk_1.nemesis-1,0,-1 do
+      if site.unk_1.nemesis[i] == nem.id then
+        site.unk_1.nemesis:erase(i)
+      end
+    end
+  end
+end
+
+function swapAdvUnit()
+
+  local newUnit
+  if args.unit then
+    newUnit = df.unit.find(tonumber(args.unit))
+  else
+    newUnit = dfhack.gui.getSelectedUnit()
+  end
+  if not newUnit then
+    print("Enter the following if you require assistance: bodyswap -help")
+    if args.unit then
+      qerror("Invalid unit id: "..args.unit)
+    else
+      qerror("Target unit not specified!")
+    end
+  end
+
+  local oldUnit = df.nemesis_record.find(df.global.ui_advmode.player_id).unit
+  if newUnit == oldUnit then
+    return
+  end
+
+  local activeUnits = df.global.world.units.active
+  local oldUnitIndex
+  if activeUnits[0] == oldUnit then
+    oldUnitIndex = 0
+  else
+    for i,u in pairs(activeUnits) do

```

Comment:
I *think* you could probably use `utils.binsearch_in_vector(activeUnits, oldUnitIndex, field)` here (but haven't tried).

Reply:
Yeah, binsearch_in_vector is the C++ name. `utils.binsearch(activeUnits, oldUnit.id, 'id')` (yeah, I screwed up my earlier example) returns `unit, true, index` if it was found. `select(3, utils.binsearch...)` would give you the index if the unit was found, but the index of another unit otherwise, so something like this might be better:
```lua
_, found, oldUnitIndex = utils.binsearch(activeUnits, oldUnit.id, 'id')
if not found then qerror('old unit not found')
-- find newUnitIndex, swap, etc.
```

Reply:
Thanks for the example!
Unless I'm messing something up, it seems that utils.binsearch isn't working reliably; it sometimes returns failure values (found == false, etc) consistently for certain cases and I have no idea why.

I'm running this in-game (in a well-populated town in adventure mode) for testing purposes. It should help illustrate my point: 
```lua
utils = require 'utils'
local activeUnits = df.global.world.units.active
for _,unit in ipairs(activeUnits) do
  print(utils.binsearch(activeUnits,unit.id,'id'))
end
```

Reply:
Hm, is it not sorted by id? I thought that was how df.unit.find() worked... can you try also checking `df.unit.find(unit.id)==unit`? Not sure if I have an adventure save handy.

Reply:
I haven't been able to reproduce this in adventure mode with ~100 units nearby

Reply:
Oh, I missed the previous reply. That's really quite strange. Mind you, failed cases have been a minority thus far; did you look through the entire output?

Reply:
The second column was all `true`s. I'll see if I can find a bigger city or something.